tests/test_mutations.py

```diff
@@ -472,40 +704,226 @@ def test_keep_mutation_parent_zero_rate(self):
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.value[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    time = rng.flat(branch_start, branch_end)
+                    site_pos = np.searchsorted([s['site'].position for s in self.sites],
+                                               position)
+                    if (site_pos == len(self.sites)
+                       or self.sites[site_pos]['site'].position != position):
+                        # new site
+                        ancestral_state = self.alleles[self.pick_allele()]
+                        site = tskit.SiteTableRow(position=position,
+                                                  ancestral_state=ancestral_state,
+                                                  metadata=b'')
+                        site_item = {'site': site,
+                                     'mutation_list': []}
+                        self.sites.insert(site_pos, site_item)
+                    else:
+                        site_item = self.sites[site_pos]
+                    mut_pos = np.searchsorted([-1 * m['time']
+                                               for m in site_item['mutation_list']],
+                                              -1 * time)
+                    mut = tskit.MutationTableRow(site=-1, node=edge.child,
+                                                 derived_state="Y",
+                                                 parent=tskit.NULL, metadata=b'')
+                    mutation_item = {'time': time,
+                                     'mutation': mut,
+                                     'new': True}
+                    site_item['mutation_list'].insert(mut_pos, mutation_item)
+                    # Sample from this to duplicate the behaviour in the C implementation
+                    rng.uniform_int(1)
+                index += 1
+                left = right
+
+    def pick_allele(self):
```

Comment:
I see, so this is just choosing an allele according to the probabilities associated with each one. If we're using numpy for randomness here, might be more obvious to use [choice](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.random.choice.html#numpy.random.choice), i.e.

```python
return np.random.choice(np.arange(len(self.alphabet_distribution), p=self.alphabet_distribution)
```

Reply:
But, I think we should probably use the GSL rng if we want to be able to compare this one-to-one with the C version (which would be nice, but might no be practical)

Reply:
yep, sorry - haven't got to this yet